File: app/src/main/java/net/ginteam/carmen/network/api/service/CompanyService.java
package net.ginteam.carmen.network.api.service;

import net.ginteam.carmen.model.ResponseModel;
import net.ginteam.carmen.model.company.CompanyModel;
import net.ginteam.carmen.model.company.MapCompanyModel;
import net.ginteam.carmen.network.api.ApiLinks;

import java.util.List;

import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * Created by Eugene on 12/27/16.
 */

public interface CompanyService {

    @GET(ApiLinks.CATALOG.COMPANIES_BY_CATEGORY)
    Observable<ResponseModel<List<CompanyModel>>> fetchCompanies(
            @Path(ApiLinks.CATALOG.ID) int categoryId,
            @Query(ApiLinks.CATALOG.LAT) String lat,
            @Query(ApiLinks.CATALOG.LNG) String lng,
            @Query(ApiLinks.CATALOG.SEARCH) String filter,
            @Query(ApiLinks.CATALOG.SORT_FIELD) String sortField,
            @Query(ApiLinks.CATALOG.SORT_TYPE) String sortType,
            @Query(ApiLinks.CATALOG.PAGE) int page
    );

    @GET(ApiLinks.CATALOG.COMPANIES_BY_BOUNDS)
    Observable<ResponseModel<List<MapCompanyModel>>> fetchCompanies(
            @Path(ApiLinks.CATALOG.ID) int categoryId,
            @Query(ApiLinks.CATALOG.SEARCH) String filter,
            @Query(ApiLinks.CATALOG.BOUNDS) String bounds
    );

    @GET(ApiLinks.CATALOG.COMPANY_BY_ID)
    Observable<ResponseModel<CompanyModel>> fetchCompanyDetail(
            @Path(ApiLinks.CATALOG.ID) int companyId,
            @Query(ApiLinks.CATALOG.LAT) String lat,
            @Query(ApiLinks.CATALOG.LNG) String lng,
            @Query(ApiLinks.CATALOG.WITH) String relations
    );

    @GET(ApiLinks.CATALOG.POPULAR_COMPANIES)
    Observable<ResponseModel<List<CompanyModel>>> fetchPopular(
            @Path(ApiLinks.CATALOG.CITY_ID) int cityId,
            @Query(ApiLinks.CATALOG.LAT) String lat,
            @Query(ApiLinks.CATALOG.LNG) String lng
    );

    @GET(ApiLinks.AUTH.GET_FAVORITES)
    Observable<ResponseModel<List<CompanyModel>>> fetchFavorites(
            @Query(ApiLinks.CATALOG.LAT) String lat,
            @Query(ApiLinks.CATALOG.LNG) String lng
    );

    @GET(ApiLinks.AUTH.GET_RECENTLY_WATCHED)
    Observable<ResponseModel<List<CompanyModel>>> fetchRecentlyWatched(
            @Query(ApiLinks.CATALOG.LAT) String lat,
            @Query(ApiLinks.CATALOG.LNG) String lng
    );

    @POST(ApiLinks.AUTH.FAVORITES_BY_ID)
    Observable<ResponseModel<String>> addToFavorites(@Path(ApiLinks.AUTH.ID) int id);

    @DELETE(ApiLinks.AUTH.FAVORITES_BY_ID)
    Observable<ResponseModel<String>> removeFromFavorites(@Path(ApiLinks.AUTH.ID) int id);


}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/api/service/CompanyService.kt
package net.ginteam.carmen.kotlin.api.service

import net.ginteam.carmen.kotlin.api.ApiSettings
import net.ginteam.carmen.kotlin.manager.ApiManager
import net.ginteam.carmen.kotlin.model.CompanyModel
import net.ginteam.carmen.kotlin.model.ResponseModel
import retrofit2.http.*
import rx.Observable

/**
 * Created by eugene_shcherbinock on 2/9/17.
 */
interface CompanyService {

    @GET(ApiSettings.Catalog.GET_COMPANIES)
    fun fetchCompanies(
            @Path(ApiSettings.Catalog.Params.ID) categoryId: Int,
            @Query(ApiSettings.Catalog.Params.PAGE) paginationPage: Int,
            @Query(ApiSettings.Catalog.Params.ORDER_BY) sortField: String,
            @Query(ApiSettings.Catalog.Params.SORTED_BY) sortType: String,
            @Query(ApiSettings.Catalog.Params.SEARCH) filter: String,
            @Query(ApiSettings.Catalog.Params.SEARCH_FIELD_TYPE) searchField: String = "name:ilike;city_id:=;rating:=;price_rel:="
    ): Observable <ResponseModel <MutableList <CompanyModel>>>

    @GET(ApiSettings.Catalog.GET_COMPANIES)
    fun fetchCompaniesForMap(
            @Path(ApiSettings.Catalog.Params.ID) categoryId: Int,
            @Query(ApiSettings.Catalog.Params.BOUNDS) bounds: String,
            @Query(ApiSettings.Catalog.Params.SEARCH) filter: String,
            @Query(ApiSettings.Catalog.Params.SEARCH_FIELD_TYPE) searchField: String = "name:ilike;city_id:=;rating:=;price_rel:="
    ): Observable <ResponseModel <MutableList <CompanyModel>>>

    @GET(ApiSettings.Catalog.GET_COMPANY_BY_ID)
    fun fetchCompany(
            @Path(ApiSettings.Catalog.Params.ID) companyId: Int,
            @Query(ApiSettings.Catalog.Relations.WITH) relations: String
    ): Observable <ResponseModel <CompanyModel>>

    @GET(ApiSettings.Catalog.GET_POPULAR_COMPANIES_BY_CITY_ID)
    fun fetchPopularCompanies(
            @Path(ApiSettings.Catalog.Params.CITY_ID) cityId: Int
    ): Observable <ResponseModel <List <CompanyModel>>>

    /*
        User companies
     */

    @GET(ApiSettings.Auth.GET_USER_FAVORITE_COMPANIES)
    fun fetchUserFavoriteCompanies(): Observable <ResponseModel <MutableList <CompanyModel>>>

    @POST(ApiSettings.Auth.GET_FAVORITE_COMPANY_BY_ID)
    fun addUserFavoriteCompany(
            @Path(ApiSettings.Auth.Params.ID) companyId: Int
    ): Observable <ResponseModel <String>>

    @DELETE(ApiSettings.Auth.GET_FAVORITE_COMPANY_BY_ID)
    fun removeUserFavoriteCompany(
            @Path(ApiSettings.Auth.Params.ID) companyId: Int
    ): Observable <ResponseModel <String>>

    @GET(ApiSettings.Auth.GET_USER_RECENTLY_WATCHED)
    fun fetchUserRecentlyWatched(): Observable <ResponseModel <List <CompanyModel>>>

    companion object {
        fun create(): CompanyService = ApiManager.retrofit.create(CompanyService::class.java)
    }

}